### PR TITLE
new: fix hgignore for real

### PR DESCRIPTION
**Pull request by sid0**
There was an attempt to fix hgignore in #4158, but unfortunately the fix
was incorrect -- hgignore glob syntax is not in fact identical to
gitignore syntax.

To get rooted ignores in Mercurial we must use the regex-based syntax,
so we have no choice but to define a separate set of ignores.

### CODE REVIEW

**Comment by sid0** [comment]
(In case it helps, I've written a gitignore to hgignore translator. It's surprisingly tricky to get right.)

File: src/cargo/ops/cargo_new.rs [hunk 1]
```diff
@@ -390,9 +390,16 @@ fn mk(config: &Config, opts: &MkOptions) -> CargoResult<()> {
...
     let cfg = global_config(config)?;
+    // Please ensure that ignore and hgignore are in sync.
     let ignore = ["/target/\n", "**/*.rs.bk\n",
         if !opts.bin { "Cargo.lock\n" } else { "" }]
         .concat();
+    // Mercurial glob ignores can't be rooted, so just sticking a 'syntax: glob' at the top of the
+    // file will exclude too much. Instead, use regexp-based ignores. See 'hg help ignore' for
+    // more.
+    let hgignore = ["^target/\n", "glob:*.rs.bk\n",
+        if !opts.bin { "^Cargo.lock\n" } else { "" }]
```
Comment (sid0) [inline src/cargo/ops/cargo_new.rs]: Do we want to root `Cargo.lock` files? I did here, but we can easily drop that too.

**Comment by alexcrichton** [comment]
@bors: r+

Thanks!

**Comment by sid0** [comment]
@alexcrichton any idea what happened to this? This might have been affected by the GH outage going on...

**Comment by alexcrichton** [comment]
@bors: r+

hmm..

**Pull request merged**